### ISSUE TITLE
Deprecate `torch.nn.functional.MultiHeadAttention` and `torch.nn.Transformer`-related modules 

### ISSUE THREAD

**Issue by mikaylagawarecki** [pytorch/pytorch#122660]
### 🚀 The feature, motivation and pitch

This issue is WIP and is a placeholder to track discussion around the deprecation of `torch.nn.functional.MultiHeadAttention` and `torch.nn.Transformer`-related `torch.nn` modules.

We observe that users would like transformer blocks to be performant and customizable. While these layers may provide an "out of the box" solution, they are monolithic modules.  Further, we note component-wise tools such as [`torch.nn.functional.scaled_dot_product_attention` ](https://www.google.com/search?q=scaled_dot_product_attention) (SDPA), [`torch.nn.attention.bias`](https://pytorch.org/docs/main/nn.attention.bias.html#module-torch.nn.attention.bias) and [`torch.nested`](https://pytorch.org/docs/main/nested.html#module-torch.nested) exist to address the issues of customizability, composability and performance. 

While there have been efforts to improve the performance and customizability of the `Transformer`/`MHA` modules, such as the addition of new masks and "fast paths" in these modules gated by configuration checks, we note that there have also been [several issues surrounding these modules](https://github.com/pytorch/pytorch/labels/oncall%3A%20transformer%2Fmha).

Examples of issues around these modules include
- Confusion around various arguments
     -  https://github.com/pytorch/pytorch/issues/118972
     - Definition of `attn_mask` is flipped compared to SDPA/implementation in other frameworks https://github.com/pytorch/pytorch/pull/120668
     - https://github.com/pytorch/pytorch/issues/121193

- Performance related issues
    -  https://github.com/pytorch/pytorch/issues/116175
    - https://github.com/pytorch/pytorch/issues/94428
    
- Bugs related to the "fast path"
   -  https://github.com/pytorch/pytorch/issues/116546
   - https://github.com/pytorch/pytorch/issues/107084
   

We also recognize that if these modules are to be deprecated, we would need to provide guidance to users on how to create a drop-in replacement using the aforementioned composable building blocks. To that end, @jbschlosser has been prototyping this with nested tensors, compile and SDPA <@jbschlosser to share more details here> :) )

**Comment by jisaacso** [comment]
cc @Chillee @mikekgfb
